### PR TITLE
[#62] [iOS] [Integrate] As a user, I can submit my answers

### LINKED ISSUE #62: [iOS] [Integrate] As a user, I can submit my answers
# Why

A user must be able to submit their answers when finished with all the questions.

# Acceptance Criteria

- Call the endpoint which is implemented in #13 to submit answers.
- Show the `Thank you` screen (#58)

### PR DESCRIPTION
**Pull request by phongvhd93**
- Close #62 

## What happened 👀

- Add new function `submitAnswer` to submit user's answer to API
- Write UT

## Insight 📝

Allow user to submit their answer to the server

## Proof Of Work 📹

Screenshot 2023-05-24 at 13 31 24

### CODE REVIEW

**Comment by markgravity** [comment]
@phongvhd93 please check CI 🙏

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/module/SurveyQuestionViewModel.kt [hunk 1]
```diff
@@ -41,4 +52,60 @@ class SurveyQuestionViewModel : BaseViewModel() {
             )
         }
     }
+
+    fun submitAnswer(questions: List<QuestionUiModel>) {
+        val surveyId = survey?.id ?: return
+        val surveySubmission = SurveySubmission(
```
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/module/SurveyQuestionViewModel.kt]: We should create a mapper extension.
```Kotlin
fun QuestionUiModel.toSurveySubmission()
```

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/uimodel/QuestionUiModel.kt [hunk 2]
```diff
@@ -32,3 +34,14 @@ fun Question.toQuestionUiModel(index: Int, totalStep: Int): QuestionUiModel = Qu
     question = this,
     step = "${index + 1}/${totalStep}"
 )
+
+fun List<QuestionUiModel>.toQuestionSubmissions(): List<QuestionSubmission> {
+    return this.map { question ->
+        QuestionSubmission(
+            id = question.id,
+            answers = question.answers.map { answer ->
```
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/uimodel/QuestionUiModel.kt]: @phongvhd93 could you please check if we should map `answers` or `userInputs` here? 🤔

Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/uimodel/QuestionUiModel.kt]: solved in 8b0c043

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/module/SurveyQuestionViewModel.kt (hunk 1 above)
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/module/SurveyQuestionViewModel.kt]: solved in 8b0c043

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/uimodel/QuestionUiModel.kt [hunk 3]
```diff
@@ -32,3 +35,17 @@ fun Question.toQuestionUiModel(index: Int, totalStep: Int): QuestionUiModel = Qu
     question = this,
     step = "${index + 1}/${totalStep}"
 )
+
+fun List<QuestionUiModel>.mapToQuestionSubmissions(): List<QuestionSubmission> {
```
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/uimodel/QuestionUiModel.kt]: I think we don't need the `map` prefix here.
```suggestion
fun List<QuestionUiModel>.toQuestionSubmissions(): List<QuestionSubmission> {
```

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/uimodel/SurveyAnswerUiModel.kt [hunk 4]
```diff
@@ -15,3 +16,9 @@ data class SurveyAnswerUiModel(
         placeholder = answer.placeholder
     )
 }
+
+fun SurveyAnswerUiModel.toAnswerSubmission() = AnswerSubmission(id, text)
+
+fun List<SurveyAnswerUiModel>.mapToAnswerSubmissions(): List<AnswerSubmission> {
```
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/uimodel/SurveyAnswerUiModel.kt]: Same here. `map` is redundant IMO.
```suggestion
fun List<SurveyAnswerUiModel>.toAnswerSubmissions(): List<AnswerSubmission> {
```

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/uimodel/QuestionUiModel.kt [hunk 5]
```diff
@@ -32,3 +35,17 @@ fun Question.toQuestionUiModel(index: Int, totalStep: Int): QuestionUiModel = Qu
...
+    return this.map { question ->
+        QuestionSubmission(
+            id = question.id,
+            answers = question.answers.mapToAnswerSubmissions()
```
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/uimodel/QuestionUiModel.kt]: Should we map `question.answers` or `question.userInputs` here?

Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/uimodel/QuestionUiModel.kt]: I made a mistake; it should be `question.userInputs.toAnswerSubmission()`. Thank you for pointing out.

Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/uimodel/QuestionUiModel.kt]: solved in 0318aa5

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/uimodel/SurveyAnswerUiModel.kt (hunk 4 above)
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/uimodel/SurveyAnswerUiModel.kt]: solved in 0318aa5

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/uimodel/QuestionUiModel.kt (hunk 3 above)
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/presentation/uimodel/QuestionUiModel.kt]: solved in 0318aa5

**Comment by AVI5HEK** [comment]
We need to fix some detekt warnings. The number exceeded the threshold. ⚠️

**Comment by phongvhd93** [comment]
@AVI5HEK, I fixed the threshold by excluding the function name of tests

**Review by AVI5HEK — APPROVED** [review]

**Review by suho — APPROVED** [review]

**Pull request merged**